File: pt5/1171.remove-zero-sum-consecutive-nodes-from-linked-list.java
import java.util.*;

class ListNode {
    int val;
    ListNode next;

    ListNode() {}

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    public ListNode removeZeroSumSublists(ListNode head) {
        ListNode dummy = new ListNode(0, head);
        Map<Integer, ListNode> prefixToNode = new HashMap<>();
        int prefix = 0;

        for (ListNode node = dummy; node != null; node = node.next) {
            prefix += node.val;
            prefixToNode.put(prefix, node);
        }

        prefix = 0;
        for (ListNode node = dummy; node != null; node = node.next) {
            prefix += node.val;
            node.next = prefixToNode.get(prefix).next;
        }

        return dummy.next;
    }
}

File: pt5/1171.remove-zero-sum-consecutive-nodes-from-linked-list.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func removeZeroSumSublists(head *ListNode) *ListNode {
	dummy := &ListNode{Next: head}
	prefixToNode := map[int]*ListNode{}
	prefix := 0

	for node := dummy; node != nil; node = node.Next {
		prefix += node.Val
		prefixToNode[prefix] = node
	}

	prefix = 0
	for node := dummy; node != nil; node = node.Next {
		prefix += node.Val
		node.Next = prefixToNode[prefix].Next
	}

	return dummy.Next
}

